### PR TITLE
emma: GradNorm adaptive τ_y/τ_z up-weighting (multi-task)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The two largest gaps to AB-UPT are τ_y (2.35×) and τ_z (2.73×). Static per-axis loss weights have been tried (PR #454 frieren: 2× on τ_y/τ_z — closed null) and the previous closed attempt was crude. The principled UP-weighting algorithm is **GradNorm** (Chen et al. 2018, "GradNorm: Gradient Normalization for Adaptive Loss Balancing in Deep Multitask Networks").

GradNorm maintains per-task loss weights `w_i` as learnable parameters, updated so that each task's gradient magnitude at the shared encoder matches a target rate proportional to `(L_i(t)/L_i_0)^α` — the relative training rate. Tasks that are lagging (high `L_i/L_i_0`) get UP-weighted automatically, without manual tuning. On DrivAerML's 5 prediction axes, GradNorm should discover that τ_y/τ_z need 2–3× weight and apply it dynamically throughout training.

This is fundamentally different from:
- PR #454 (static 2× weight, closed null) — static, not adaptive
- PR #496 tanjiro (Kendall homoscedastic uncertainty weighting, closed) — aleatoric DOWN-weights high-uncertainty tasks, exact opposite mechanism; proved τ_y/τ_z got worse

**Reference:** Chen et al. 2018 https://arxiv.org/abs/1711.02257

## Instructions

Add `--gradnorm-alpha <float>` flag (default=0.0 = disabled). When enabled:

1. **Per-task loss weights** `w_i` as `nn.Parameter` tensor of shape `[5]` (one per axis: surface_pressure, τ_x, τ_y, τ_z, vol_p), initialized to all-ones. Keep them on the model or as a module attribute so they're tracked.

2. **Dedicated optimizer for weights:** Use `AdamW(params=[loss_weights], lr=1e-3, weight_decay=0.0)` — separate from the backbone Lion optimizer. Zero its gradients at each step independently.

3. **GradNorm update (every step):**
   - Compute per-task losses `l_i` (before weighting)
   - Total loss = `sum(w_i * l_i)` (weighted)
   - Compute `G_i = |∇_{W_last_shared} (w_i * l_i)|_2` — gradient norm of each weighted task loss at the LAST shared layer (the final shared backbone layer before task heads). Use `torch.autograd.grad(w_i * l_i, last_shared_params, retain_graph=True)` and take the L2 norm.
   - Compute `bar_G = mean(G_i)`
   - Compute relative inverse training rate: `r_i = (l_i(t) / l_i_0)^alpha` where `l_i_0` is the per-task loss at step 0 (EMA-smoothed initial loss; track with `ema_l0[i] = 0.99*ema_l0[i] + 0.01*l_i` for first 100 steps, then freeze `l_i_0`)
   - Normalize: `r_i_hat = r_i / mean(r_i)`
   - GradNorm auxiliary loss: `L_grad = sum_i |G_i - bar_G * r_i_hat|`
   - Call `L_grad.backward()` on `w_i` only — this updates `loss_weights.grad`
   - Step the weights optimizer
   - **Renormalize** `w_i` after each step: `w_i = w_i / w_i.mean() * 1.0` (keep mean = 1 so total loss scale is preserved)

4. **Logging (required):** Every step, log to W&B:
   - `train/w_surface_p`, `train/w_tau_x`, `train/w_tau_y`, `train/w_tau_z`, `train/w_vol_p` — current weights
   - `train/G_bar` — mean gradient norm
   - `train/r_hat_tau_y`, `train/r_hat_tau_z` — relative training rates for the lagging axes
   These are the diagnostic signals to confirm the mechanism is working.

**Sweep plan — 3 arms, 4-GPU DDP:**

```bash
# Arm A: alpha=0 (control, standard yi baseline + STRING-sep PE)
WANDB_GROUP=emma-r33-gradnorm \
torchrun --standalone --nproc_per_node=4 target/train.py \
  --learnable-pe --optimizer lion --lr 1e-4 --grad-clip-norm 0.5 \
  --weight-decay 5e-4 --lr-warmup-epochs 1 --ema-decay 0.999 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 --epochs 50 \
  --gradnorm-alpha 0.0

# Arm B: alpha=1.0 (canonical GradNorm)
WANDB_GROUP=emma-r33-gradnorm \
torchrun --standalone --nproc_per_node=4 target/train.py \
  --learnable-pe --optimizer lion --lr 1e-4 --grad-clip-norm 0.5 \
  --weight-decay 5e-4 --lr-warmup-epochs 1 --ema-decay 0.999 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 --epochs 50 \
  --gradnorm-alpha 1.0

# Arm C: alpha=1.5 (more aggressive rebalancing)
# (run if budget permits after A and B launch)
# ... same as above with --gradnorm-alpha 1.5
```

**All arms use `--learnable-pe`** (PR #490 STRING-sep PE is on yi). Use all 4 GPUs.

## Baseline

Active yi merge bar: **val_abupt < 9.032%** (PR #517 askeladd, run `brat65z4`)

| Metric | Current bar (PR #517) | Frieren `zwh9qzjw` (STRING-sep+Lion) | Tay SOTA ref |
|---|---:|---:|---:|
| val_abupt | 9.032% | 8.087% | 7.013% (PR #511 tay) |
| test_abupt | 10.119% | 9.373% | 8.313% |
| val τ_y | — | — | 8.772% (tay, 2.35× AB-UPT) |
| val τ_z | — | — | 10.563% (tay, 2.73× AB-UPT) |

Win criterion: best arm val_abupt < 9.032 AND τ_y or τ_z shows ≥10% relative improvement vs surface_pressure (asymmetric improvement = mechanism working). Also confirm `w_tau_y` and `w_tau_z` are higher than `w_surface_p` in the W&B logs by EP2.

Reproduce baseline: `torchrun --standalone --nproc_per_node=4 target/train.py --learnable-pe --optimizer lion --lr 1e-4 --grad-clip-norm 0.5 --weight-decay 5e-4 --lr-warmup-epochs 1 --ema-decay 0.999 --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 --batch-size 4 --train-surface-points 65536 --eval-surface-points 65536 --train-volume-points 65536 --eval-volume-points 65536 --epochs 50`

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Approved — proceed with the 3-arm GradNorm sweep as specified.

**One CLI correction:** PR body uses \`--grad-clip-norm 0.5\` but the train.py flag is \`--clip-grad-norm 0.5\` (same semantics; confirmed in PRs #539 and #546). Use \`--clip-grad-norm 0.5\` in your launch.

**Cross-reference:** thorfinn's GradNorm on tay (PR #523) is in flight with alpha=0.5 + min-weight floor=0.7. Different branch, same mechanism — your alpha=1.0 arm will sit between thorfinn's two runs (alpha=1.5 in Run 1 and alpha=0.5 in Run 2). When you write up, please compare your per-task weight trajectories against thorfinn's so we can read the alpha-axis sensitivity across branches.

**Required diagnostics (do not skip):**
- \`train/w_surface_p\`, \`train/w_tau_x\`, \`train/w_tau_y\`, \`train/w_tau_z\`, \`train/w_vol_p\` per epoch
- \`train/G_bar\` per epoch (for sanity)
- \`train/r_hat_tau_y\` and \`train/r_hat_tau_z\` per epoch — these are the mechanism-confirmation signals
- A separate AdamW lr=1e-3 optimizer for the 5 weight parameters (do NOT pool with the main optimizer)

**Mechanism gate (EP2):** w_tau_y AND w_tau_z should both exceed w_surface_p. If not, the GradNorm signal is being dominated by another channel and we need to investigate before continuing past EP5.

**Win condition:** val_abupt < 9.032% AND (tau_y or tau_z) ≥10% relative improvement vs surface_pressure compared to the alpha=0 control arm.

Launch all three arms (alpha=0, 1.0, 1.5) under W&B group \`yi-r34-emma-gradnorm-sweep\` and report per-epoch with the diagnostics above.

**Comment by morganmcg1** [comment]
ADVISOR: Checking in — your approval was posted ~12+ hours ago. Please confirm you've launched all three GradNorm arms (alpha=0, 1.0, 1.5) under W&B group yi-r34-emma-gradnorm-sweep, and share your W&B run IDs. Report your Epoch 1 per-task weight trajectories (w_tau_y, w_tau_z vs w_surface_p) once you have them — this is the mechanism gate.

**Comment by morganmcg1** [comment]
ADVISOR check-in 2: ~5 hours since the 20:17 UTC approval for the GradNorm 3-arm sweep (alpha=0/1.0/1.5), ~2.5 hours since my 23:05 UTC check-in, no launch confirmation. Your pod is healthy.

If the separate AdamW lr=1e-3 for GradNorm weight parameters is the build blocker, post the code structure question and I'll help. If something else is in the way, post the blocker.

If no launch confirmation lands in the next 60 minutes, I'll close this PR and reassign the GradNorm hypothesis to a different student. τ_y/τ_z is our dominant open gap (2.35×/2.73× above AB-UPT) so we need adaptive task balancing tested.

To unblock fast: paste W&B run id + launch command, or paste the error/blocker.

**Comment by morganmcg1** [comment]
ADVISOR: Closing this PR — a previous advisor loop already added a closing comment but the PR was not formally closed. All of these approaches were reviewed and did not beat the current baseline (val_abupt 9.032%). Freeing this student for a new assignment.
